src/OTCNexus.sol

```diff
@@ -0,0 +1,283 @@
+
+        makerRfs[msg.sender].push(idToRfs[rfsIdCounter]);
+
+        emit RfsCreated(
+            rfsIdCounter,
+            msg.sender,
+            _token0,
+            _tokensAccepted,
+            _amount0,
+            _usdPrice,
+            _priceMultiplier,
+            _deadline
+        );
+
+        return rfsIdCounter++;
+    }
+
+    // //TODO TO BE UPDATED WITH NEW LOGIC
+    // // /**
+    // //  * @notice take an RFS with deposit
+    // //  * @param _id ID of the RFS
+    // //  * @param _amount1 amount of token1
+    // //  * @return success
+    // //  */
+    // function takeFixedRfs(
+    //     uint256 _id,
+    //     uint256 _amount1,
+    //     uint256 index
+    // ) external returns (bool success) {
+    //     RFS memory rfs = getRfs(_id);
+
+    //     // sanity checks
+    //     if (rfs.removed || rfs.token0 == address(0)) revert Router__RfsRemoved();
+    //     if (_amount1 == 0) revert Router__InvalidTokenAmount();
+    //     if (_amount1 > rfs.amount1) revert Router__Amount1TooHigh();
+
+    //     //Maybe we can just switch and ask directly for approve instead of checking the allowance.
+    //     if (IERC20(rfs.tokensAccepted[index]).allowance(msg.sender, address(this)) < _amount1)
+    //         revert Router__AllowanceToken1TooLow();
+
+    //     // transfer token1 to the maker
+    //     success = IERC20(rfs.tokensAccepted[index]).transferFrom(msg.sender, rfs.maker, _amount1);
+    //     if (!success) revert Router__TransferToken1Failed();
+
+    //     // compute _amount0 based on _amount1 and implied quote, and update RFS amounts
+    //     uint256 _amount0 = OtcMath.getTakerAmount0(rfs.amount0, rfs.amount1, _amount1);
+
+    //     // transfer token0 to the taker
+    //     success = IERC20(rfs.token0).transfer(msg.sender, _amount0);
+    //     if (!success) revert Router__TransferToken0Failed();
+
+    //     // update RFS
+    //     // todo use function updateRfs and emit RfsUpdated event
+    //     rfs.amount0 -= _amount0;
+    //     rfs.amountsToPay[index] -= _amount1;
+    //     if (rfs.amount0 == 0 || rfs.amountsToPay[index] == 0) {
+    //         rfs.removed = true;
+    //         emit RfsRemoved(_id);
+    //     }
+    //     idToRfs[_id] = rfs;
+
+    //     emit RfsFilled(_id, msg.sender, _amount0, _amount1);
+    // }
+
+    /**
+     * @notice remove an RFS
+     * @param _id ID of the RFS
+     * @param _permanentlyDelete if the RFS should be permanently deleted
+     * @return success
+     */
+    function removeRfs(uint256 _id, bool _permanentlyDelete) external returns (bool success) {
+        RFS memory rfs = idToRfs[_id];
+        // only the maker can remove the RFS
+        if (msg.sender != rfs.maker) revert Router__NotMaker();
+
+        if (rfs.isDeposited) {
+            // refund deposited tokens to the maker
+            success = IERC20(rfs.token0).transfer(rfs.maker, rfs.amount0);
+            if (!success) revert Router__RefundToken0Failed();
```

Comment:
The revert already throws errors that can be seen on the blockchain :). Custom errors are just a way to use explainable errors, instead of using strings, that are really costly.

Reply:
Ok, my reasoning was - all errors could have the level SERIOUS but this one if happens could be marked as FATAL because sth really wrong happened. For example if swap fails but no tokens were deposit (ie. approval approach) nothing really affects user (maybe only gas to try operation). But if this one fails it means the user has lost his coin.

Ok, so I see Solidity does not have the mechanism I know from other software. So be it. :)

Reply:
It doesn't mean he lost. Maybe there' sjust a problem. The contract will just revert the execution and nothing will happen to the RFS.

Reply:
It might be just another problem, but my understanding is - once the user has his coin deposited he can get back either swapped coins or got transfer back with function removeRfs (ie. when rfs is deleted). If removeRfs fails on transfer function it mean a big problem occurred.

Reply:
Yes, but it shouldn't happen in any case. (If the code is secure :D)